### PR TITLE
 Management and definition of Week Phases

### PR DESCRIPTION
**Pull request by AndreaSR95**
Link on youtrack: https://polito-se2-21-09.myjetbrains.com/youtrack/issue/S202109SPG-22

Implementation on server side of (under services folder):

- A WeekPhaseDate Class  which purpose is to represent a specific instant of time inside a GENERIC week. I didn't use dayjs here, because it requires also the definition of the month and year, which we don't use.

- A WeekPhaseInterval Class which purpose is to rappresent an interval of time inside a week. The class provides also an handler function (to be executed when this interval of time becomes active). Handlers are currently empty and have to be defined depending on the specs given.

- A WeekPhasesService service which purpose is:

1. set as active the correct/current WeekPhase
2. set a cron clock that checks every minute if the phase is to be changed
3. if the phase is changed, change the active phase and run the new phase handler
4. give the possibility to take control over the phases overriding the active phase and disabling cron
5. give the possibility to check the current phase

PS for the reviewer: can't run the tests on the cron function.

### CODE REVIEW

File: server/app.js [hunk 1]
```diff
@@ -5,6 +5,12 @@ var logger = require("morgan");
 var cors = require("cors");
 var dao = require("./dao/dao");
 
+// ----------
+// /timeManager
+// ----------
+const { TimeManager } = require("./models/timeManager");
+var timeManagerObj = new TimeManager();
```
Comment (MarcoBramini) [inline server/app.js]: Call this WeekPhaseService. I don't know why a service is in the models folder. Move all the code you wrote in the service folder, under a subfolder named WeekPhaseService. The main file should have the same name as the subfolder.

Also, this service has to work as the dao. We need it to be single instance for testing. Copy the dao initialization and get rid of the class.

File: server/models/timeManager.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+const dayjs = require("dayjs");
+
+
+//Handle of phase called only when phase is enabled
+
+class TimeManager {
+    constructor(){
+        this.phaseList = this.getPhaseList();
+        this.activePhase = this.getTrueActivePhase();
+        this.activePhase.handler();
```
Comment (MarcoBramini) [inline server/models/timeManager.js]: let's decide the cron if it's time to run the handler. This can lead to error (and I'm pretty sure is a big bug)

File: server/models/timeManager.js [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    getTrueActivePhase(){
+        let nowjs = dayjs();
+        let now = new Time(nowjs.day(), nowjs.hour(), nowjs.minute(), nowjs.second());
+
+        for(let phase of this.phaseList){
+            if(phase.isActiveDuringData(now)){
+                return phase;
+            }
+        }
+        return undefined;
```
Comment (MarcoBramini) [inline server/models/timeManager.js]: Undefined must not be returned by a method. Is better to return null.

File: server/models/timeManager.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        return [
+            //ID, StartTime, EndTime, Description, Handler
+            new WeekPhase("ID1", new Time("Monday","00","00","00"), new Time("Saturday","19","30","00"), "description bla bla bla", ()=>{console.log("try handle 1");}),
+            new WeekPhase("ID2", new Time("Saturday","19","30","00"), new Time("Monday","00","00","00"), "description bla bla bla", ()=>{console.log("try handle 2");})
+        ];
+    }
+
+    getTrueActivePhase(){
+        let nowjs = dayjs();
+        let now = new Time(nowjs.day(), nowjs.hour(), nowjs.minute(), nowjs.second());
```
Comment (MarcoBramini) [inline server/models/timeManager.js]: We don't need the seconds, as I already said. Remove them

File: server/models/timeManager.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+                return;
+            this.phaseIDOverride = null;
+            this.enableCron();
+            return;
+        }
+        //this means we are setting the time to "virtual"
+        if(phaseID != this.phaseIDOverride){
+            this.phaseIDOverride = phaseID;
+            this.disableCron();
+            this.phaseList.filter((phase)=> phaseID==phase.ID )[0].handler();
```
Comment (MarcoBramini) [inline server/models/timeManager.js]:
```suggestion
            this.phaseList.filter((phase)=> phaseID===phase.ID )[0].handler();
```

File: server/models/timeManager.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+        //this means we are setting the time to "not virtual"
+        if(phaseID == null){ 
+            if(this.phaseIDOverride == null)
+                return;
+            this.phaseIDOverride = null;
+            this.enableCron();
+            return;
+        }
+        //this means we are setting the time to "virtual"
+        if(phaseID != this.phaseIDOverride){
```
Comment (MarcoBramini) [inline server/models/timeManager.js]:
```suggestion
        if(phaseID !== this.phaseIDOverride){
```

File: server/models/timeManager.js [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+            this.phaseIDOverride = phaseID;
+            this.disableCron();
+            this.phaseList.filter((phase)=> phaseID==phase.ID )[0].handler();
+        }
+    }
+
+    enableCron(){
+        this.cronTask = cron.schedule('* * * * *', ()=> {
```
Comment (MarcoBramini) [inline server/models/timeManager.js]: The cron must be initialized one time only. Move the schedule method in a module variable. (like the MongoDB client in the dao )

File: server/models/timeManager.js [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+            console.log("Checking if new phase...");
+        
+            let phase = this.getTrueActivePhase();
+            if(phase.ID != this.activePhase.ID){
```
Comment (MarcoBramini) [inline server/models/timeManager.js]:
```suggestion
            if(phase.ID !== this.activePhase.ID){
```

File: server/models/timeManager.js [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+                console.log("new phase detected!");
+                this.activePhase = phase;
+                this.activePhase.handler();
+            }
+            console.log("No new phase");
+        });
+        this.cronTask.start();
```
Comment (MarcoBramini) [inline server/models/timeManager.js]: This line must be the only one in this method. This makes this method useless. Call the cron start directly where you need to.

File: server/models/timeManager.js [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    disableCron(){
+        this.cronTask.stop();
```
Comment (MarcoBramini) [inline server/models/timeManager.js]: Same as above, this method is useless

File: server/models/timeManager.js [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    getCurrentPhaseID(){
+        if(this.phaseIDOverride != null)
```
Comment (MarcoBramini) [inline server/models/timeManager.js]:
```suggestion
        if(this.phaseIDOverride !== null)
```

File: server/models/timeModel.js [hunk 12]
```diff
@@ -0,0 +1,43 @@
+class Time{
```
Comment (MarcoBramini) [inline server/models/timeModel.js]: You can't reinvent the wheel. If you want to do something like this you must use dayjs.
Call this WeekPhaseLimit or WeekPhaseBoundary or make a model with both the startTime and endTime and call it WeekPhaseInterval (better choice).

File: server/models/timeModel.js [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+    constructor(day, hour, min, sec){
+        this.day = day;
+        this.dayNumber = this.convertDayIntoNumber(day);
+        this.hour = hour;
+        this.min = min;
+        this.sec = sec;
+    }
+
+    //METHODS
+    convertDayIntoNumber(day){
```
Comment (MarcoBramini) [inline server/models/timeModel.js]: This is pretty much useless, as I said. This is not done by the user.

File: server/models/timeModel.js [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+            day = "4";
+            break;
+          case "Friday":
+            day = "5";
+            break;
+          case "Saturday":
+            day = "6";
+            break;
+          case "Sunday":
+            day = "7";
```
Comment (MarcoBramini) [inline server/models/timeModel.js]: This is a bug, Sunday has index 0.

File: server/models/timeModel.js [hunk 15]
```diff
@@ -0,0 +1,43 @@
+class Time{
+    constructor(day, hour, min, sec){
+        this.day = day;
+        this.dayNumber = this.convertDayIntoNumber(day);
+        this.hour = hour;
+        this.min = min;
+        this.sec = sec;
```
Comment (MarcoBramini) [inline server/models/timeModel.js]: This is useless and not required

File: server/models/weekPhase.js [hunk 16]
```diff
@@ -0,0 +1,41 @@
+"use strict";
+
+class WeekPhase {
+  constructor(ID, startTime, endTime, description, handler) {
+    this.ID = ID;
+    this.startTime = startTime;
+    this.endTime = endTime;
+    this.description = description || "no description";
```
Comment (MarcoBramini) [inline server/models/weekPhase.js]: Description is of no use. Remove it

File: server/models/weekPhase.js [hunk 17]
```diff
@@ -0,0 +1,41 @@
+"use strict";
+
+class WeekPhase {
+  constructor(ID, startTime, endTime, description, handler) {
+    this.ID = ID;
```
Comment (MarcoBramini) [inline server/models/weekPhase.js]: When the ID is alone, it's lowercase

File: server/models/weekPhase.js [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+    return [this.startTime, this.endTime];
+  }
+
+  isActiveDuringData(dateObj){//Returns true if the parameter time is between the start and end time of this phase
+
+    let dateString = dateObj.toString();
+    let startTimeString = this.startTime.toString();
+    let endTimeString = this.endTime.toString();
+
+    if(startTimeString < endTimeString){//phase not cross weeks
```
Comment (MarcoBramini) [inline server/models/weekPhase.js]: We don't have phases crossing weeks as constraint, remove this check and simplify this.

**Review by MarcoBramini — CHANGES_REQUESTED** [review]
The api methods setWeekPhaseOverride and getCurrentWeekPhase are not implemented.

They also need to be tested.

File: server/models/timeModel.js (hunk 12 above)
Comment (AndreaSR95) [reply server/models/timeModel.js]: The problem with jdays was that I can't define "Monday at 15:00", but I have to define "2021, November, 28, Monday at 15:00". So I thought I had 3 choices. 1) I use jdays defining a random yyyy/mm/dd and ignoring those fields in my code. 2) I find an alternative to jdays that doesn't require me to define a random values for yyyy/mm/dd. 3) I write the class with only the info needed. If 1) is considered a better choice, I'll do that.

File: server/models/timeModel.js (hunk 14 above)
Comment (AndreaSR95) [reply server/models/timeModel.js]: It's only a bug if I want to export to jdays. It's 7 instead of 0 because I made dates as comparable "numeric strings" ( <, >, =). Sunday comes after Saturday, so the string "70000" > "60000". But if we decided to use jdays, this whole file is useless and to be deleted.

File: server/models/timeManager.js (hunk 2 above)
Comment (AndreaSR95) [reply server/models/timeManager.js]: If this is about the comment, I think that comment is not written well. The handler function is ONLY called inside the function we pass to cron as parameter.

Comment (MarcoBramini) [reply server/models/timeManager.js]: Yep, I saw the condition in the cron method and it should work.

File: server/models/timeModel.js (hunk 12 above)
Comment (MarcoBramini) [reply server/models/timeModel.js]: You can create the object WeekPageInterval, which should contain:
```
- startWeekday //0,1,2,3,4,5,6
- startHour // 12
- startMinute // 30

- endWeekday
- endHour
- endMinute

```
And then define a function to check if a dayjs date is contained in the interval:
```
function isDateWithinInterval(date):
   dateWeekday = date.weekday
   dateHour = date.hour
   dateMinute = date.minute
   
   dateStr = dateWeekday+dateHour+dateMinute

   startStr = startWeekday+startHour+startMinute
   endStr = endWeekday+endHour+endMinute

   return (dateStr >= startStr && dateStr < endStr)
```

File: server/app.js [hunk 19]
```diff
@@ -20,6 +20,8 @@ var orderHandlers = require("./handlers/order");
 
 var productHandlers = require("./handlers/product");
 
+var weekPhasesService = require("./services/WeekPhsaseService/weekPhasesService");
```
Comment (MarcoBramini) [inline server/app.js]: Still a typo in the path...

Comment (MarcoBramini) [reply server/app.js]: Follow the naming convention of every other service. (weekphases_service)

File: server/services/WeekPhsaseService/weekPhaseDate.js [hunk 20]
```diff
@@ -0,0 +1,13 @@
+class WeekPhaseDate{
```
Comment (MarcoBramini) [inline server/services/WeekPhsaseService/weekPhaseDate.js]: The name is not correct, call this WeekPhaseBoundary as I already commented.

File: server/services/WeekPhsaseService/weekPhaseInterval.js [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+    this.id = id;
+    this.startTime = startTime;
+    this.endTime = endTime;
+    this.handler = handler;
+  }
+
+  //METHODS
+  //Returns true if the parameter is between the start and end time of this phase
+  isDateWithinInterval(date){
+    let dateString = date.toString();
```
Comment (MarcoBramini) [inline server/services/WeekPhsaseService/weekPhaseInterval.js]: Make this method accept as input a dayjs date. Copy the exact behavior of the comment I wrote a week ago.

File: server/services/WeekPhsaseService/weekPhasesFunctions.js [hunk 22]
```diff
@@ -0,0 +1,36 @@
+const weekPhases = require("./weekPhaseInterval");
+const dayjs = require("dayjs");
+const { WeekPhaseDate } = require("./weekPhaseDate");
+
+exports.getTrueActivePhase = ()=>{
```
Comment (MarcoBramini) [inline server/services/WeekPhsaseService/weekPhasesFunctions.js]: Move all this functions to the same service file, delete this file.

File: server/services/WeekPhsaseService/weekPhaseInterval.js [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+    let startTimeString = this.startTime.toString();
+    let endTimeString = this.endTime.toString();
+
+    return (dateString>=startTimeString && dateString<endTimeString);
+  }
+}
+
+const weekPhases = [
```
Comment (MarcoBramini) [inline server/services/WeekPhsaseService/weekPhaseInterval.js]: This should not be here but in the main service file.

File: server/services/WeekPhsaseService/weekPhasesFunctions.js [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+    let nowjs = dayjs();
+    let now = new WeekPhaseDate(nowjs.day(), nowjs.hour(), nowjs.minute());
+
+    for(let phase of weekPhases.weekPhases){
```
Comment (MarcoBramini) [inline server/services/WeekPhsaseService/weekPhasesFunctions.js]:
```suggestion
    let activePhase = weekPhases.filter(wp => wp.isDateWithinInterval(now));
    return (activePhase.length > 0) ? activePhase[0] : null;
```

File: server/services/WeekPhsaseService/weekPhasesFunctions.js [hunk 25]
```diff
@@ -0,0 +1,36 @@
+const weekPhases = require("./weekPhaseInterval");
+const dayjs = require("dayjs");
+const { WeekPhaseDate } = require("./weekPhaseDate");
+
+exports.getTrueActivePhase = ()=>{
+    let nowjs = dayjs();
+    let now = new WeekPhaseDate(nowjs.day(), nowjs.hour(), nowjs.minute());
```
Comment (MarcoBramini) [inline server/services/WeekPhsaseService/weekPhasesFunctions.js]: You should hide this in the method. This is not a boundary so this is a logic issue. Build the string as in the previous comments.
If it makes you feel better you can put the building in a function, but it would be pretty much useless.

File: server/services/WeekPhsaseService/weekPhasesService.js [hunk 26]
```diff
@@ -0,0 +1,39 @@
+"use strict"
+
+var cron = require('node-cron');
```
Comment (MarcoBramini) [inline server/services/WeekPhsaseService/weekPhasesService.js]: Importing this module multiple time will result in multiple instances of it running. Clone the dao behavior.

File: server/services/WeekPhsaseService/weekPhasesService.js [hunk 27]
```diff
@@ -0,0 +1,39 @@
...
+    activePhase = phase;
+    activePhase.handler();
+  }
+  else
+    console.log("No new phase");
+});
+
+cronTask.start();
+
+exports.getCurrentPhaseID = () => getCurrentPhaseID(activePhase, phaseIDOverride);
```
Comment (MarcoBramini) [inline server/services/WeekPhsaseService/weekPhasesService.js]: Handlers are still not implemented and there's no way to set or get the phase from the extern. 
You didn't implement the apis as required.

File: server/test/test.js [hunk 28]
```diff
@@ -1691,3 +1691,51 @@ describe("User Login API tests:", () => {
...
+
+//tests on time
+
+const weekPhasesService = require("../services/WeekPhsaseService/weekPhasesService");
+const { WeekPhaseInterval } = require("../services/WeekPhsaseService/weekPhaseInterval");
+const weekPhases = require("../services/WeekPhsaseService/weekPhaseInterval");
+const { WeekPhaseDate } = require("../services/WeekPhsaseService/weekPhaseDate");
+
+
+describe("weekPhase",()=>{
```
Comment (MarcoBramini) [inline server/test/test.js]: Stick to our testing convention. these tests are absolutely random until the edge cases are tested and everything is well described.

**Review by MarcoBramini — CHANGES_REQUESTED** [review]

**Review by MarcoBramini — APPROVED** [review]
